File: VulkanGameEngine/RenderMesh2DComponent.cpp
//#include "RenderMesh2DComponent.h"
//#include "MemoryManager.h"
//
//RenderMesh2DComponent::RenderMesh2DComponent() : GameObjectComponent()
//{
//}
//
//RenderMesh2DComponent::RenderMesh2DComponent(SharedPtr<GameObject> parentGameObjectPtr, String name, uint32 meshBufferIndex) : GameObjectComponent(this, parentGameObjectPtr, name, ComponentTypeEnum::kRenderMesh2DComponent)
//{
//	std::vector<Vertex2D> SpriteVertexList =
//	{
//	  { {0.0f, 0.5f},  {0.0f, 0.0f}, {1.0f, 0.0f, 0.0f, 1.0f} },
//	  { {0.5f, 0.5f},  {1.0f, 0.0f}, {0.0f, 1.0f, 0.0f, 1.0f} },
//	  { {0.5f, 0.0f},  {1.0f, 1.0f}, {0.0f, 0.0f, 1.0f, 1.0f} },
//	  { {0.0f, 0.0f},  {0.0f, 1.0f}, {1.0f, 1.0f, 0.0f, 1.0f} }
//	};
//	std::vector<uint32> SpriteIndexList =
//	{
//	  0, 1, 3,
//	  1, 2, 3
//	};
//
//	mesh = std::make_shared<Mesh2D>(std::make_shared<RenderMesh2DComponent>(*this), SpriteVertexList, SpriteIndexList, meshBufferIndex);
//}
//
//RenderMesh2DComponent::~RenderMesh2DComponent()
//{
//}
//
//SharedPtr<RenderMesh2DComponent> RenderMesh2DComponent::CreateRenderMesh2DComponent(SharedPtr<GameObject> parentGameObjectPtr, String name, uint32 meshBufferIndex)
//{
//	std::vector<Vertex2D> SpriteVertexList =
//	{
//	  { {0.0f, 0.5f},  {0.0f, 0.0f}, {1.0f, 0.0f, 0.0f, 1.0f} },
//	  { {0.5f, 0.5f},  {1.0f, 0.0f}, {0.0f, 1.0f, 0.0f, 1.0f} },
//	  { {0.5f, 0.0f},  {1.0f, 1.0f}, {0.0f, 0.0f, 1.0f, 1.0f} },
//	  { {0.0f, 0.0f},  {0.0f, 1.0f}, {1.0f, 1.0f, 0.0f, 1.0f} }
//	};
//	std::vector<uint32> SpriteIndexList =
//	{
//	  0, 1, 3,
//	  1, 2, 3
//	};
//
//	SharedPtr<RenderMesh2DComponent> gameObject = MemoryManager::AllocateRenderMesh2DComponent();
//	new (gameObject.get()) RenderMesh2DComponent(parentGameObjectPtr, name, meshBufferIndex);
//	return gameObject;
//}
//
//void RenderMesh2DComponent::Input(float deltaTime)
//{
//}
//
//void RenderMesh2DComponent::Update(float deltaTime)
//{
//	mesh->Update(deltaTime);
//}
//
//void RenderMesh2DComponent::BufferUpdate(VkCommandBuffer& commandBuffer, float deltaTime)
//{
//	mesh->BufferUpdate(commandBuffer, deltaTime);
//	mesh->Update(deltaTime);
//}
//
//void RenderMesh2DComponent::Draw(VkCommandBuffer& commandBuffer, VkPipeline& pipeline, VkPipelineLayout& pipelineLayout, VkDescriptorSet& descriptorSet, SceneDataBuffer& sceneProperties)
//{
//	mesh->Draw(commandBuffer, pipeline, pipelineLayout, descriptorSet, sceneProperties);
//}
//
//void RenderMesh2DComponent::Destroy()
//{
//	mesh->Destroy();
//}
//
//SharedPtr<GameObjectComponent> RenderMesh2DComponent::Clone() const
//{
//	return std::make_shared<RenderMesh2DComponent>(*this);
//}
//
//size_t RenderMesh2DComponent::GetMemorySize() const
//{
//	return sizeof(RenderMesh2DComponent);
//}

File: VulkanGameEngineLevelEditor/GameEngineAPI/RenderMesh2DComponent.cs
﻿using Silk.NET.Vulkan;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using VulkanGameEngineLevelEditor.Vulkan;

namespace VulkanGameEngineLevelEditor.GameEngineAPI
{
    public unsafe class RenderMesh2DComponent : GameObjectComponent
    {
        public Mesh2D mesh { get; protected set; } = new Mesh2D();
        public RenderMesh2DComponent() 
        { 

        }

        private void Initialize(String name, uint meshBufferIndex)
        {
            List<Vertex2D> spriteVertexList = new List<Vertex2D>
            {
                new Vertex2D(new vec2(0.0f, 0.5f), new vec2(0.0f, 0.0f), new vec4(1.0f, 0.0f, 0.0f, 1.0f)),
                new Vertex2D(new vec2(0.5f, 0.5f), new vec2(1.0f, 0.0f), new vec4(0.0f, 1.0f, 0.0f, 1.0f)),
                new Vertex2D(new vec2(0.5f, 0.0f), new vec2(1.0f, 1.0f), new vec4(0.0f, 0.0f, 1.0f, 1.0f)),
                new Vertex2D(new vec2(0.0f, 0.0f), new vec2(0.0f, 1.0f), new vec4(1.0f, 1.0f, 0.0f, 1.0f))
            };

            List<uint> spriteIndexList = new List<uint> { 0, 1, 3, 1, 2, 3 };

            mesh = new Mesh2D(meshBufferIndex);
        }

        public static RenderMesh2DComponent CreateRenderMesh2DComponent(String name, uint meshBufferIndex)
        {
            List<Vertex2D> spriteVertexList = new List<Vertex2D>
            {
                new Vertex2D(new vec2(0.0f, 0.5f), new vec2(0.0f, 0.0f), new vec4(1.0f, 0.0f, 0.0f, 1.0f)),
                new Vertex2D(new vec2(0.5f, 0.5f), new vec2(1.0f, 0.0f), new vec4(0.0f, 1.0f, 0.0f, 1.0f)),
                new Vertex2D(new vec2(0.5f, 0.0f), new vec2(1.0f, 1.0f), new vec4(0.0f, 0.0f, 1.0f, 1.0f)),
                new Vertex2D(new vec2(0.0f, 0.0f), new vec2(0.0f, 1.0f), new vec4(1.0f, 1.0f, 0.0f, 1.0f))
            };

            List<uint> spriteIndexList = new List<uint> { 0, 1, 3, 1, 2, 3 };

            RenderMesh2DComponent gameObject = MemoryManager.AllocateGameRenderMesh2DComponent();
            gameObject.Initialize(name, meshBufferIndex);
            return gameObject;
        }

        public override void Update(float deltaTime)
        {
            mesh.Update(deltaTime);
        }

        public override void Update(CommandBuffer commandBuffer, float deltaTime)
        {
            mesh.BufferUpdate(commandBuffer, deltaTime);
            mesh.Update(deltaTime);
        }

        public override void Draw(CommandBuffer commandBuffer, Pipeline pipeline, PipelineLayout shaderPipelineLayout, DescriptorSet descriptorSet, SceneDataBuffer sceneProperties)
        {
            mesh.Draw(commandBuffer, pipeline, shaderPipelineLayout, descriptorSet, sceneProperties);
        }

        public override void Destroy()
        {
            mesh.Destroy();
        }

        public override int GetMemorySize()
        {
            return sizeof(RenderMesh2DComponent);
        }
        
        public VulkanBuffer<MeshProperitiesBuffer> GetMeshPropertiesBuffer() 
        { 
            return mesh.GetMeshPropertiesBuffer(); 
        }
    }
}
